pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go

```diff
@@ -20,47 +20,135 @@ import (
+	preFilterStateKey = "PreFilter" + Name
+)
+
+// preFilterState computed at PreFilter and used at Filter.
+type preFilterState struct {
+	meta *predicates.PodAffinityMetadata
+}
+
+// Clone the prefilter state.
+func (s *preFilterState) Clone() framework.StateData {
+	copy := &preFilterState{
+		meta: s.meta.Clone(),
+	}
+	return copy
+}
 
 // Name returns name of the plugin. It is used in logs, etc.
 func (pl *InterPodAffinity) Name() string {
 	return Name
 }
 
+// PreFilter invoked at the prefilter extension point.
+func (pl *InterPodAffinity) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	var meta *predicates.PodAffinityMetadata
+	var allNodes []*nodeinfo.NodeInfo
+	var havePodsWithAffinityNodes []*nodeinfo.NodeInfo
+	var err error
+	if allNodes, err = pl.snapshotSharedLister.NodeInfos().List(); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("failed to list NodeInfos: %v", err))
+	}
+	if havePodsWithAffinityNodes, err = pl.snapshotSharedLister.NodeInfos().HavePodsWithAffinityList(); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("failed to list NodeInfos with pods with affinity: %v", err))
+	}
+	if meta, err = predicates.GetPodAffinityMetadata(pod, allNodes, havePodsWithAffinityNodes); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("Error calculating podAffinityMetadata: %v", err))
+	}
+
+	s := &preFilterState{
+		meta: meta,
+	}
+	cycleState.Write(preFilterStateKey, s)
+
+	return nil
+}
+
+// PreFilterExtensions returns prefilter extensions, pod add and remove.
+func (pl *InterPodAffinity) PreFilterExtensions() framework.PreFilterExtensions {
+	return pl
+}
+
+// AddPod from pre-computed data in cycleState.
+func (pl *InterPodAffinity) AddPod(ctx context.Context, cycleState *framework.CycleState, podToSchedule *v1.Pod, podToAdd *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	meta, err := getPodAffinityMetadata(cycleState)
+	if err != nil || meta == nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	meta.UpdateWithPod(podToAdd, podToSchedule, nodeInfo.Node(), 1)
+	return nil
+}
+
+// RemovePod from pre-computed data in cycleState.
+func (pl *InterPodAffinity) RemovePod(ctx context.Context, cycleState *framework.CycleState, podToSchedule *v1.Pod, podToRemove *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	meta, err := getPodAffinityMetadata(cycleState)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	meta.UpdateWithPod(podToRemove, podToSchedule, nodeInfo.Node(), -1)
+	return nil
+}
+
+func getPodAffinityMetadata(cycleState *framework.CycleState) (*predicates.PodAffinityMetadata, error) {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if s, ok := c.(*preFilterState); ok {
+		return s.meta, nil
+	}
+	return nil, fmt.Errorf("%+v  convert to interpodaffinity.state error", c)
```

Comment:
It would be better to return the error earlier

```go
	if s, ok := c.(*preFilterState); !ok {
		return nil, fmt.Errorf("%+v  convert to interpodaffinity.state error", c)
	}
	return s.meta, nil
```

Reply:
this does not compile, s is only defined inside the if statement.

Reply:
```go
        s, ok := c.(*preFilterState)
	if  !ok {
		return nil, fmt.Errorf("%+v  convert to interpodaffinity.state error", c)
	}
	return s.meta, nil
```

Reply:
and why is that better?

Reply:
> and why is that better?

FYI: https://github.com/golang/go/wiki/CodeReviewComments#in-band-errors

> And encourages more robust and readable code:

```
value, ok := Lookup(key)
if !ok  {
    return fmt.Errorf("no value for %q", key)
}
return Parse(value)

Reply:
> > and why is that better?
> 
> FYI: https://github.com/golang/go/wiki/CodeReviewComments#in-band-errors
> 
> > And encourages more robust and readable code:
> 
> ```
> value, ok := Lookup(key)
> if !ok  {
>     return fmt.Errorf("no value for %q", key)
> }
> return Parse(value)
> ```

The paragraph you linked to talks about a completely different thing, it talks about in-band errors (i.e., embedding the error as a special value in the return type). 

The quote "And encourages more robust and readable code:" is not about whether to return the error in the if or else clause, it is about golang's ability to return multiple values compared to C/C++, and hence allowing to explicitly return an error type instead of embedding it in ```value```.

I will change it if you feel strongly about this, but again the link above is irrelevant.

Reply:
The section below does talk about error indentation though: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
Thanks, return error earlier is always preferred in error handling IMO